### ISSUE TITLE
Thread running after request in app engine

### ISSUE THREAD

**Issue by ubaldo** [segmentio/analytics-python#34]
In google app engine I consistently get the following error (not always, but very often). See stack trace below

The effect is that the event doesn't get recorded. When I first encounter this, I tried to isolate the call to analytics.track() in a task (using the handy deferred library), to avoid side effects on a user facing http request. So, I sleep better but still many events don't get recorded 

I'm only interested in executing one analytics track per task in the task queue. Was looking at the code to try avoiding the starting a new thread etc (threading is very tricky in app engine), but I don't see how to avoid it, even it queue-size is 1 I don't think it would work. 

Ideas?

2014-11-06 10:39:41.179 consumer is running...
D 2014-11-06 10:39:41.179 enqueued track.
D 2014-11-06 10:39:41.181 making request: {"batch": [{"event": "Payment Failed", "anonymousId": null, "context": {"library": {"version": "1.0.3", "name": "analytics-python"}},
I 2014-11-06 10:39:41.195 Starting new HTTPS connection (1): api.segment.io
D 2014-11-06 10:39:41.836 "POST /v1/batch HTTP/1.1" 200 None
D 2014-11-06 10:39:41.838 data uploaded successfully
D 2014-11-06 10:39:41.959 successfully flushed 1 items.
E 2014-11-06 10:49:39.608 Thread running after request. Creation traceback: File "/base/data/home/runtimes/python27/python27_lib/versions/1/google/appengine/runtime/runtime.p
I 2014-11-06 10:49:39.609 This request caused a new process to be started for your application, and thus caused your application code to be loaded for the first time. This requ
W 2014-11-06 10:49:39.609 Threads started by this request continued executing past the hard deadline

**Comment by suhaasprasad** [comment]
Any update on this? I'm having the same problem and it's preventing me from doing anything with Segment.

**Comment by jmgirven** [comment]
I would also be interested in an update.

**Comment by sperand-io** [comment]
@f2prateek can we look at this together next week? Thanks!

**Comment by Jonovono** [comment]
Any update on this?

**Comment by thinq4yourself** [comment]
:+1: bump... Any update on this?

**Comment by david0673** [comment]
any update?

**Comment by jgershen** [comment]
Looking at this old bug -- is this still an issue after the 1.2.5 release? We fixed some bugs around threading that might have contributed.

**Comment by cdman** [comment]
Yes, this is still a problem with the latest version as of now (version 1.2.5).

To recapitulate:
- Google Appengine does not support background threads for frontend instances (the instance type you generally want to use)
- Starting (and not stopping) threads on GAE will make your request "hang" because it requires each thread started during the request to finish before it finishes

To use the library on GAE, you need to:
- manually install it, its dependencies (`requests` and `python-dateutil`) and the transitive dependencies of `python-dateutil` (`six`) - [vendoring](https://cloud.google.com/appengine/docs/python/tools/using-libraries-python-27) may be of some help
- you also need to include `requests-toolbelt` to make requests work (or activate the paid-only sockets feature) - see [Issuing HTTP(S) Requests](https://cloud.google.com/appengine/docs/python/issue-requests) - you may still see [some spurious warnings](https://github.com/sigmavirus24/requests-toolbelt/issues/163)
- after each request you need to call `.join()`, otherwise your request will hang. Ie:

```python
analytics.track(...
analytics.join()
```

A much better architecture for this library would be to make both the "queuing" and "fetching" part pluggable. In particular for GAE Python 2.7 you probably want to:

- use a push queue for queuing (with a dedicated worker pulling from the queue and forwarding the times)
- use the urlfetch service (or perhaps even the asynchronous urlfetch service)

**Comment by cdman** [comment]
Ok, to work on GAE, you need to monkeypatch it like this:

```python
import analytics.client
import analytics.request
_original_analytics_enqueue = analytics.client.Client._enqueue
def _monkeypatched_analytics_enqueue(client, message):
    if client.queue.put != analytics.client.queue.Queue.put:
        def monkeypatched_analytics_queue_put(msg, block):
            analytics.request.post(constants.SEGMENT_KEY, batch=[msg])
        client.queue.put = monkeypatched_analytics_queue_put

    if client.consumer.running:
        client.consumer.running = False

    _original_analytics_enqueue(client, message)
analytics.client.Client._enqueue = _monkeypatched_analytics_enqueue
```

This forces all the messages to be sent synchronously and stops any running extra threads to prevent GAE from locking up by waiting on those threads.

**Comment by f2prateek** [comment]
Merging into https://github.com/segmentio/analytics-python/issues/101

**Comment by mdornseif** [comment]
You can get usable results on google app engine by manually flushingrushing and exiting the thread. General idea:

    analytics.identify(self.credential.uid, context)
    … do your stuff …
    analytics.flush()
    analytics.default_client.consumer.pause()  # this EXITS the thread

**Comment by savraj** [comment]
about to add segment to a google app engine project. going to try mdornseif's idea.
